src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+        /// owned by a thread.
+        /// </summary>
+        private class LockEntry
+        {
+            [ThreadStatic]
+            private static LockEntry t_lockEntryHead;
+
+            private long _lockID;
+            private LockEntry _next;
+            public ushort _readerLevel;
+
+            private LockEntry(long lockID)
+            {
+                _lockID = lockID;
+            }
+
+            public bool HasLockID(long lockID)
+            {
+                return _lockID == lockID;
+            }
+
+            public bool IsEmpty
+            {
+                get
+                {
+                    return _readerLevel == 0;
+                }
+            }
+
+            private static void VerifyNoNonemptyEntryInListAfter(long lockID, LockEntry afterEntry)
+            {
+                Debug.Assert(lockID != 0);
+                Debug.Assert(afterEntry != null);
+
+#if DEBUG
+                for (LockEntry currentEntry = afterEntry._next; currentEntry != null; currentEntry = currentEntry._next)
+                {
+                    Debug.Assert(currentEntry._lockID != lockID || currentEntry.IsEmpty);
+                }
+#endif
+            }
+
+            public static LockEntry GetCurrent(long lockID)
+            {
+                Debug.Assert(lockID != 0);
+
+                LockEntry headEntry = t_lockEntryHead;
+                for (LockEntry currentEntry = headEntry; currentEntry != null; currentEntry = currentEntry._next)
+                {
+                    if (currentEntry._lockID == lockID)
+                    {
+                        VerifyNoNonemptyEntryInListAfter(lockID, currentEntry);
+
+                        // The lock ID assignment is only relevant when the entry is not empty, since the lock ID is not reset
+                        // when its state becomes empty. Empty entries can be poached by other ReaderWriterLocks.
+                        return currentEntry.IsEmpty ? null : currentEntry;
+                    }
+                }
+                return null;
+            }
+
+            public static LockEntry GetOrCreateCurrent(long lockID)
+            {
+                Debug.Assert(lockID != 0);
+
+                LockEntry headEntry = t_lockEntryHead;
+                if (headEntry != null)
+                {
+                    if (headEntry._lockID == lockID)
+                    {
+                        VerifyNoNonemptyEntryInListAfter(lockID, headEntry);
+                        return headEntry;
+                    }
+
+                    if (headEntry.IsEmpty)
+                    {
+                        VerifyNoNonemptyEntryInListAfter(lockID, headEntry);
+                        headEntry._lockID = lockID;
+                        return headEntry;
+                    }
+                }
+
+                return GetOrCreateCurrentSlow(lockID, headEntry);
```

Comment:
Nit: what's the purpose of the separated-out Slow function?  I'm used to seeing this when it's likely the rest of the method would get inlined, but I wouldn't expect that to happen in this case.  If it's not about perf, maybe a different name rather than "Slow".